app/open_ai_client.rb

```diff
@@ -1,111 +1,227 @@
+    end
+  end
+
+  def create_grading_function
+    assign_grades = ENV.fetch("ASSIGN_GRADES", "false") == "true"
+    if assign_grades
+      {
+        type: "function",
+        function: {
+          name: "create_grading",
+          description: "Creates grading for a student submission",
+          parameters: {
+            type: "object",
+            properties: {
+              status: {
+                type: "string",
+                enum: ["accepted", "rejected"]
+              },
+              feedback: {
+                type: "string",
+                description: "The feedback to be added to a student submission"
+              },
+              grades: {
+                type: "array",
+                items: {
+                  type: "object",
+                  properties: {
+                    evaluationCriterionId: {
+                      type: "string",
+                      enum: Submission.new.evaluation_criteria_ids,
+                      description: "The Id of evaluation criteria"
+                    },
+                    grade: {
+                      type: "integer",
+                      description: "The grade value choosen from allowed grades array for choosen evaluatuion_criterion_id for a submission based on quality"
+                    }
+                  },
+                  required: ["evaluationCriterionId", "grade"]
+                }
+              }
+            },
+            required: ["status", "feedback", "grades"]
+          }
+        }
+      }
+    else
+      {
+        type: "function",
+        function: {
+          name: "create_grading",
+          description: "Creates grading for a student submission",
+          parameters: {
+            type: "object",
+            properties: {
+              status: {
+                type: "string",
+                enum: ["accepted", "rejected"]
+              },
+              feedback: {
+                type: "string",
+                description: "The feedback to be added to a student submission"
+              }
+            },
+            required: ["feedback"]
+          }
+        }
+      }
+    end
+  end
+
+  def function
+    if ENV.fetch("SKIP_GRADING", "false") == "true"
+      create_feedback_function
+    else
+      create_grading_function
+    end
+  end
+
   def ask
     puts prompt
     response = @client.chat(
-        parameters: {
-            model: @model,
-            messages: [
-                { role: "system", content: prompt }
-            ],
-            temperature: @temperature,
-        })
+      parameters: {
+        model: @model,
+        messages: [
+          {role: "system", content: prompt}
+        ],
+        tools: [function],
+        temperature: @temperature
+      }
+    )
     puts response
-    response.dig("choices", 0, "message", "content")
+    response.dig("choices", 0, "message", "tool_calls", 0, "function", "arguments")
```

Comment:
Function call may not happen always (I have seen this quite many times) also why do we need to follow this structure?

Also for handling response lets go with the example mentioned in the readme; Lets make it explicit on which function is getting called; 

https://github.com/alexrudall/ruby-openai?tab=readme-ov-file#functions